File: arcgis_map_sdk_android/android/src/main/kotlin/dev/fluttercommunity/arcgis_map_sdk_android/CustomLocationProvider.kt
package dev.fluttercommunity.arcgis_map_sdk_android

import com.arcgismaps.location.CustomLocationDataSource
import com.arcgismaps.location.Location
import dev.fluttercommunity.arcgis_map_sdk_android.model.UserPosition
import dev.fluttercommunity.arcgis_map_sdk_android.model.toAGSPoint
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.asSharedFlow

class CustomLocationProvider : CustomLocationDataSource.LocationProvider {
    private val _locations = MutableSharedFlow<Location>(replay = 1)
    private val _headings = MutableSharedFlow<Double>(replay = 1)

    override val locations: Flow<Location> = _locations.asSharedFlow()
    override val headings: Flow<Double> = _headings.asSharedFlow()

    fun updateLocation(position: UserPosition) {
        _locations.tryEmit(
            Location.create(
                position = position.latLng.toAGSPoint(),
                horizontalAccuracy = position.accuracy ?: 0.0,
                verticalAccuracy = position.accuracy ?: 0.0,
                speed = position.velocity ?: 0.0,
                course = position.heading ?: 0.0,
                lastKnown = false,
            )
        )
        position.heading?.let {
            _headings.tryEmit(it)
        }
    }
}

File: arcgis_map_sdk_ios/ios/arcgis_map_sdk_ios/Sources/arcgis_map_sdk_ios/CustomLocationProvider.swift
//
//  Created by Julian Bissekkou on 27.11.23.
//

import Foundation
import ArcGIS

final class CustomLocationProvider: LocationProvider {
    private var locationContinuation: AsyncStream<Location>.Continuation?
    private var headingContinuation: AsyncStream<Double>.Continuation?
    
    
    // Exposed stream
    var locations: AsyncStream<Location> {
        AsyncStream { @Sendable continuation in
            self.locationContinuation = continuation
            continuation.onTermination = { _ in
                self.locationContinuation = nil
            }
        }
    }
    
    var headings: AsyncStream<Double> {
        AsyncStream { continuation in
            self.headingContinuation = continuation
            continuation.onTermination = { @Sendable _ in
                self.headingContinuation = nil
            }
        }
    }
    
    // Push location from outside
    public func setNewLocation(_ position: UserPosition) {
        let loc = Location(
            position: position.latLng.toAGSPoint(),
            horizontalAccuracy: position.accuracy ?? 0,
            verticalAccuracy: position.accuracy ?? 0,
            speed: position.velocity ?? 0,
            course: position.heading ?? 0,
            isLastKnown: false
        )
        locationContinuation?.yield(loc)
        if let heading = position.heading {
            headingContinuation?.yield(heading)
        }
    }
}

